lib/event-subscribers.loader.ts

```diff
@@ -31,39 +45,126 @@ export class EventSubscribersLoader
     );
     if (!eventListenerMetadata) {
       return;
     }
+
     const { event, options } = eventListenerMetadata;
-    const listenerMethod = !!options?.prependListener
+    const listenerMethod = this.registerListenerMethodBasedOn(options);
+
+    if (isRequestScoped) {
+      this.registerRequestScopedListener({
+        event,
+        eventListenerInstance: instance,
+        listenerMethod,
+        listenerMethodKey: methodKey,
+        moduleRef,
+        options,
+      });
+    } else {
+      listenerMethod(
+        event,
+        (...args: unknown[]) => instance[methodKey].call(instance, ...args),
+        options,
+      );
+    }
+  }
+
+  private registerListenerMethodBasedOn(options?: OnEventOptions) {
+    return Boolean(options?.prependListener)
       ? this.eventEmitter.prependListener.bind(this.eventEmitter)
       : this.eventEmitter.on.bind(this.eventEmitter);
+  }
+
+  private async registerRequestScopedListener(eventListenerContext: {
+    listenerMethod: EventEmitter2['on'];
+    event: string | symbol | (string | symbol)[];
+    eventListenerInstance: Record<string, any>;
+    moduleRef: Module;
+    listenerMethodKey: string;
+    options?: OnEventOptions;
+  }) {
+    const {
+      listenerMethod,
+      event,
+      eventListenerInstance,
+      moduleRef,
+      listenerMethodKey,
+      options,
+    } = eventListenerContext;
 
     listenerMethod(
       event,
-      (...args: unknown[]) => instance[methodKey].call(instance, ...args),
+      async (...args: unknown[]) => {
+        const contextId = ContextIdFactory.create();
+
+        this.registerEventPayloadByContextId(args, contextId);
+
+        const contextInstance = await this.injector.loadPerContext(
+          eventListenerInstance,
+          moduleRef,
+          moduleRef.providers,
+          contextId,
+        );
+        return contextInstance[listenerMethodKey].call(
+          contextInstance,
+          ...args,
+        );
+      },
       options,
     );
   }
+
+  private registerEventPayloadByContextId(
+    eventPayload: unknown[],
+    contextId: ContextId,
+  ) {
+    /*
+      **Required explanation for the ternary below**
+
+      We need the conditional below because an event can be emitted with a variable amount of arguments.
+      For instance, we can do `this.eventEmitter.emit('event', 'payload1', 'payload2', ..., 'payloadN');` 
+      
+      All payload arguments are internally stored as an array. So, imagine we emitted an event as follows:
+
+      `this.eventEmitter.emit('event', 'payload');
+      
+      if we registered the original `eventPayload`, when we try to inject it in a listener, it'll be retrieved as [`event`].
+      However, whoever is using this library would certainly expect the event payload to be a single object, not an array,
+      since this is what we emitted above.
+    */
+
+    const payloadObjectOrArray =
+      eventPayload.length > 1 ? eventPayload : eventPayload[0];
```

Comment:
I'll double check this, but I think the `eventPayload` here will actually always be an array, even if we passed a string. That's because the `eventPayload` here is actually an alias for the function's `args` parameters, which is a rest parameter (and thus, an array)

Reply:
I added a test for the case you mentioned in [this commit](https://github.com/nestjs/event-emitter/pull/544/commits/afeeb0c134822655c20097ad9daa1d5b132d0a3b) and it works as expected, therefore I think we don't have to check against an array here